SignalServiceKit/src/Messages/OWSReadReceiptManager.m

```diff
@@ -0,0 +1,191 @@
+
+// A map of "thread unique id"-to-"read receipt" for read receipts that
+// we will send to our linked devices.
+//
+// Should only be accessed while synchronized on the OWSReadReceiptManager.
+@property (nonatomic, readonly) NSMutableDictionary<NSString *, OWSReadReceipt *> *toLinkedDevicesReadReceiptMap;
+
+// Should only be accessed while synchronized on the OWSReadReceiptManager.
+@property (nonatomic) BOOL isProcessing;
+
+@end
+
+#pragma mark -
+
+@implementation OWSReadReceiptManager
+
++ (instancetype)sharedManager
+{
+    static OWSReadReceiptManager *sharedMyManager = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedMyManager = [[self alloc] initDefault];
+    });
+    return sharedMyManager;
+}
+
+- (instancetype)initDefault
+{
+    OWSMessageSender *messageSender = [TextSecureKitEnv sharedEnv].messageSender;
+
+    return [self initWithMessageSender:messageSender];
+}
+
+- (instancetype)initWithMessageSender:(OWSMessageSender *)messageSender
+{
+    self = [super init];
+
+    if (!self) {
+        return self;
+    }
+
+    _messageSender = messageSender;
+
+    _toLinkedDevicesReadReceiptMap = [NSMutableDictionary new];
+
+    OWSSingletonAssert();
+
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(databaseViewRegistrationComplete)
+                                                 name:kNSNotificationName_DatabaseViewRegistrationComplete
+                                               object:nil];
+
+    return self;
+}
+
+- (void)dealloc
+{
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)databaseViewRegistrationComplete
+{
+    [self scheduleProcessing];
+}
+
+// Schedules a processing pass, unless one is already scheduled.
+- (void)scheduleProcessing
+{
+    DispatchMainThreadSafe(^{
+        @synchronized(self)
+        {
+            if ([TSDatabaseView hasPendingViewRegistrations]) {
+                return;
+            }
+            if (self.isProcessing) {
+                return;
+            }
+
+            self.isProcessing = YES;
+
+            // Process read receipts every N seconds.
+            //
+            // We want a value high enough to allow us to effectively deduplicate,
+            // read receipts without being so high that we risk not sending read
+            // receipts due to app exit.
+            const CGFloat kProcessingFrequencySeconds = 3.f;
+            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(kProcessingFrequencySeconds * NSEC_PER_SEC)),
+                dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0),
+                ^{
+                    [self process];
+                });
+        }
+    });
+}
+
+- (void)process
+{
+    @synchronized(self)
+    {
+        self.isProcessing = NO;
+
+        NSArray<OWSReadReceipt *> *readReceiptsToSend = [[self.toLinkedDevicesReadReceiptMap allValues] copy];
```

Comment:
nit: seems like this `copy` isn't doing anything.

Reply:
Removed.